[JIT] Add may_contains_alias function in SchemaInfo class

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81655
* #81653
* #81562
* #81561
* #81575
* #81518
* __->__ #81444
* #81439

- Created may_contain_alias method in SchemaInfo which is a wrapper around FunctionSchema may_contain_alias that also accounts for argument values. This is done using similar logic to AliasDB using an internal understanding of wildcard sets and container object
- Added a multitude of tests for various graph edge cases (inputs aliasing, outputs aliasing, multiple input wildcards, multiple container objects, etc...).

Commit:
[JIT] Add may_contains_alias function in SchemaInfo class

[ghstack-poisoned]

Commit:
Update on "[JIT] Add may_contains_alias function in SchemaInfo class"


- Created may_contain_alias method in SchemaInfo which is a wrapper around FunctionSchema may_contain_alias that also accounts for argument values. This is done using similar logic to AliasDB using an internal understanding of wildcard sets and container object
- Added a multitude of tests for various graph edge cases (inputs aliasing, outputs aliasing, multiple input wildcards, multiple container objects, etc...).

[ghstack-poisoned]

davidberard98 commented on test/cpp/jit/test_schema_info.cpp:
```diff
@@ -254,6 +254,106 @@ TEST(FunctionSchemaMayContainAliasTest, InputAndOutputContainers) {
       {c10::SchemaArgType::input, 0}, {c10::SchemaArgType::output, 0}, false));
 }
 
+TEST(SchemaInfoMayContainAliasTest, ContainAliasInputsEqual) {
```
nice tests!

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -170,6 +170,36 @@ bool SchemaInfo::may_alias(
+bool SchemaInfo::mayContainAliasImpl(
+    const c10::SchemaArgument& lhs,
+    const c10::SchemaArgument& rhs) {
+  bool types_can_alias = schema_.canAliasTypeSetsAlias(
+      schema_.getAliasTypeSetContainedTypes(schema_.mapTypeToAliasTypeSet(
+          schema_.getCorrectList(lhs.type)[lhs.index].type())),
+      schema_.mapTypeToAliasTypeSet(
+          schema_.getCorrectList(rhs.type)[rhs.index].type()));
```
could you add some intermediate variables here to make this easier to read?

davidberard98 replied:
I know you already documented this below but can you add a quick comment on why this is different from the FunctionSchema implementation?

(you could also just add a comment directing readers to the other documentation, search for `Note [` to see examples of how it's often done in pytorch codebase)

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -264,6 +300,24 @@ void SchemaInfo::generateAliasMaps() {
 
+  // Fills wildcard_set with container created wildcards.
+  // ie if one argument is a container and it contains an element that aliases
+  // the other argument, then the second argument is added to the wildcard set.
```
an example would be nice to help understand this! You can probably just copy one out of the tests. (I was a bit confused until I read the tests)

davidberard98 commented on torch/csrc/utils/schema_info.cpp:
```diff
@@ -214,12 +244,12 @@ void SchemaInfo::initSchemaInfo() {
         for (size_t i = 0; i < arguments_list.size(); i++) {
-          if (arguments_list[i].alias_info()) {
-            if (arguments_list[i].alias_info()->isWildcardAfter()) {
+          const c10::Argument argument = arguments_list[i];
```
nit: `const c10::Argument&`

davidberard98 commented on torch/csrc/utils/schema_info.h:
```diff
@@ -39,6 +39,11 @@ struct TORCH_API SchemaInfo {
       const c10::SchemaArgument& lhs,
       const c10::SchemaArgument& rhs);
 
+  bool may_contain_alias(
```
can you document the bidirectional behavior here as well?

Commit:
Update on "[JIT] Add may_contains_alias function in SchemaInfo class"


- Created may_contain_alias method in SchemaInfo which is a wrapper around FunctionSchema may_contain_alias that also accounts for argument values. This is done using similar logic to AliasDB using an internal understanding of wildcard sets and container object
- Added a multitude of tests for various graph edge cases (inputs aliasing, outputs aliasing, multiple input wildcards, multiple container objects, etc...).

[ghstack-poisoned]

Commit:
Update on "[JIT] Add may_contains_alias function in SchemaInfo class"


- Created may_contain_alias method in SchemaInfo which is a wrapper around FunctionSchema may_contain_alias that also accounts for argument values. This is done using similar logic to AliasDB using an internal understanding of wildcard sets and container object
- Added a multitude of tests for various graph edge cases (inputs aliasing, outputs aliasing, multiple input wildcards, multiple container objects, etc...).

[ghstack-poisoned]

Commit:
Update on "[JIT] Add may_contains_alias function in SchemaInfo class"


- Created may_contain_alias method in SchemaInfo which is a wrapper around FunctionSchema may_contain_alias that also accounts for argument values. This is done using similar logic to AliasDB using an internal understanding of wildcard sets and container object
- Added a multitude of tests for various graph edge cases (inputs aliasing, outputs aliasing, multiple input wildcards, multiple container objects, etc...).

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
